_posts/2016-10-31-angular2-discussion-portal.md

```diff
@@ -0,0 +1,141 @@
+        this.subscribeToData();
+    });
+}
+
+private subscribeToData(): void {
+    this.timerSubscription = Observable.timer(5000).first().subscribe(() => this.refreshData());
+}
+{% endhighlight %}
+
+This function is responsible for refreshing comments in real time. When someone add a new comment everyone who is reading this post will see it without need for refreshing page.
+
+Firstly, we invoke `commentService.getComments(id: number)` to retrieve comments from the backend. On each success call we run `subscribeToData()` which after 5s invoke once again `refreshData()`.
+It ensures that data on UI are constantly refresh.
+
+You may wonder why we invoke `subscribeToData()` after successful call instead of adding in `ngOnInit()` method:
+{% highlight typescript %}
+this.timerSubscription = Observable.timer(5000).subscribe(() => this.refreshData());
+{% endhighlight %}
+
+This code call `refreshData()` after every 5s. The problem occurs when on backend side processing a request takes more than 5s. Then the new request (send after 5s) will begin to queue leading to clogging server.
+Invoking `refreshData()` after successful call pretends such behaviour. We have certainty that only one request in time is sent to a backend service.
+
+The same architecture is used in `DashboardComponent` for refreshing dashboard data (all available posts):
+
+{% highlight typescript %}
+private refreshData(): void {
+    this.postsSubscription = this.postService.getPosts().subscribe(posts => {
+        this.posts = posts;
+        this.subscribeToData();
+    });
+}
+
+private subscribeToData(): void {
+    this.timerSubscription = Observable.timer(5000).first().subscribe(() => this.refreshData());
+}
+{% endhighlight %}
+
+Remember that you should unsubscribe from all subscriptions in `ngOnDestroy()` method to prevent memory leaks.
+
+{% highlight typescript %}
+public ngOnDestroy(): void {
+    if (this.postsSubscription) {
+        this.postsSubscription.unsubscribe();
+    }
+    if (this.timerSubscription) {
+        this.timerSubscription.unsubscribe();
+    }
+}
+{% endhighlight %}
+
+### 3 Directories structure
+While writing first application in angular2 I was wondering what is the best practise for files structure in the repository. During working with angular2 I worked out structure which is simple and easy to maintain while adding new functionalities.
+
+<img src="/images/blog/posts/angular2-discussion-portal/packages_structure.png"/>
+
+I have three main directories: 
+
+* html - to keep all html files, divided by pages and components. When the new component on dashboard is created new file is added to `html\dashboard`.
+* scss - all scss files. Lack of division by components, because basically each page has one scss file.
+* ts - typescripts files, divided by pages and type of classes (components/services/models). I also have `shared` directory which contains all code shared between different pages.
+
+
+### Sum  up
```

Comment:
You could mention in summary about different approach to having live data on the page. Instead of making request every 5 seconds to get new data you can use websockets and server can push messages to the user browser so thanks to that when many users will have open page they won't be hitting server every 5s. Instead just server will send new message to user when new messaged was published.

__publish-subscribe messaging system examples:__

Ruby: 
* Faye https://faye.jcoglan.com/
* Rails ActionCable http://edgeguides.rubyonrails.org/action_cable_overview.html

Elixir/Phoenix: 
* http://www.phoenixframework.org/docs/channels